pdearena/models/pderefiner.py

```diff
@@ -0,0 +1,422 @@
+
+        elif dataloader_idx == 1:
+            if self._mode == "1D" or self._mode == "2D":
+                loss_vecs = self.compute_rolloutloss(batch)
+            else:
+                raise NotImplementedError(f"{self._mode}")
+            # summing across "time axis"
+            loss_mse = loss_vecs["mse"].sum()
+            loss_mse_t = loss_vecs["mse"].cumsum(0)
+            chan_avg_loss = loss_mse / \
+                (self.pde.n_scalar_components + self.pde.n_vector_components)
+            self.log("valid/unrolled_loss", loss_mse)
+            return {
+                "unrolled_loss": loss_mse,
+                "loss_timesteps": loss_mse_t,
+                "unrolled_chan_avg_loss": chan_avg_loss,
+                "corr": loss_vecs["corr"]
+            }
+
+    def test_epoch_end(self, outputs: List[Any]):
+        assert len(outputs) > 1
+        if len(outputs[0]) > 0:
+            for key in outputs[0][0].keys():
+                if "loss" in key:
+                    loss_vec = torch.stack([outputs[0][i][key]
+                                           for i in range(len(outputs[0]))])
+                    mean, std = utils.bootstrap(loss_vec, 64, 1)
+                    self.log(f"test/{key}_mean", mean)
+                    self.log(f"test/{key}_std", std)
+        if len(outputs[1]) > 0:
+            unrolled_loss = torch.stack(
+                [outputs[1][i]["unrolled_loss"] for i in range(len(outputs[1]))])
+            loss_timesteps_B = torch.stack(
+                [outputs[1][i]["loss_timesteps"] for i in range(len(outputs[1]))])
+            loss_timesteps = loss_timesteps_B.mean(0)
+            log_timesteps = range(loss_timesteps.shape[0], max(
+                1, loss_timesteps.shape[0] // 10))
+            for i in log_timesteps:
+                self.log(f"test/intime_{i}_loss", loss_timesteps[i])
+
+            mean, std = utils.bootstrap(unrolled_loss, 64, 1)
+            self.log("test/unrolled_loss_mean", mean)
+            self.log("test/unrolled_loss_std", std)
+
+            # Correlation
+            corr_timesteps_B = torch.stack(
+                [outputs[1][i]["corr"] for i in range(len(outputs[1]))], dim=0
+            )
+            corr_timesteps = corr_timesteps_B.mean(0)
+            for threshold in [0.8, 0.9, 0.95]:
+                self.log(
+                    f"tests/time_till_corr_lower_{threshold}", (corr_timesteps > threshold).float().sum())
+            for t in log_timesteps:
+                self.log(f"tests/corr_at_{t}", corr_timesteps[t])
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(
+            self.model.parameters(), lr=self.hparams.lr)
+        return optimizer
+
+    def on_fit_start(self):
+        self.ema.register()
+
+    def on_train_batch_end(self, *args, **kwargs):
+        self.ema.update()
+
+    def on_validation_start(self):
+        self.apply_ema()
+
+    def on_validation_end(self):
+        self.remove_ema()
+
```

Comment:
Should also be done for `on_test_start` and `on_test_end`?

Reply:
Good point. Usually we test after having finished the training with a learning rate cooldown, which makes the EMA basically identical to the original weights. But to be consistent, let's also go with applying the EMA during testing.